plugins/inputs/nfsclient/nfsclient.go

```diff
@@ -0,0 +1,326 @@
+	"GETACL",
+	"SETACL",
+	"FS_LOCATIONS",
+	"RELEASE_LOCKOWNER",
+	"SECINFO",
+	"FSID_PRESENT",
+	"EXCHANGE_ID",
+	"CREATE_SESSION",
+	"DESTROY_SESSION",
+	"SEQUENCE",
+	"GET_LEASE_TIME",
+	"RECLAIM_COMPLETE",
+	"LAYOUTGET",
+	"GETDEVICEINFO",
+	"LAYOUTCOMMIT",
+	"LAYOUTRETURN",
+	"SECINFO_NO_NAME",
+	"TEST_STATEID",
+	"FREE_STATEID",
+	"GETDEVICELIST",
+	"BIND_CONN_TO_SESSION",
+	"DESTROY_CLIENTID",
+	"SEEK",
+	"ALLOCATE",
+	"DEALLOCATE",
+	"LAYOUTSTATS",
+	"CLONE",
+}
+
+var nfsopFields = []string{
+	"ops",
+	"trans",
+	"timeouts",
+	"bytes_sent",
+	"bytes_recv",
+	"queue_time",
+	"response_time",
+	"total_time",
+}
+
+func convert(line []string) []int64 {
+	var nline []int64
+	for _, l := range line[1:] {
+		f, _ := strconv.ParseInt(l, 10, 64)
+		nline = append(nline, f)
+	}
+	return nline
+}
+
+func in(list []string, val string) bool {
+	for _, v := range list {
+		if v == val {
+			return true
+		}
+	}
+	return false
+}
+
+func (n *NFSClient) parseStat(mountpoint string, export string, version string, line []string, fullstat bool, acc telegraf.Accumulator) error {
+	tags := map[string]string{"mountpoint": mountpoint, "serverexport": export}
+	nline := convert(line)
+	first := strings.Replace(line[0], ":", "", 1)
+
+	var fields = make(map[string]interface{})
+
+	if fullstat && first == "events" && len(nline) >= len(eventsFields) {
+		for i, t := range eventsFields {
+			fields[t] = nline[i]
+		}
+		acc.AddFields("nfs_events", fields, tags)
+	} else if fullstat && first == "bytes" && len(nline) >= len(bytesFields) {
+		for i, t := range bytesFields {
+			fields[t] = nline[i]
+		}
+		acc.AddFields("nfs_bytes", fields, tags)
+	} else if fullstat && first == "xprt" && len(line) > 1 {
+		switch line[1] {
+		case "tcp":
+			if len(nline)+2 >= len(xprttcpFields) {
+				for i, t := range xprttcpFields {
+					fields[t] = nline[i+2]
+				}
+				acc.AddFields("nfs_xprttcp", fields, tags)
+			}
+		case "udp":
+			if len(nline)+2 >= len(xprtudpFields) {
+				for i, t := range xprtudpFields {
+					fields[t] = nline[i+2]
+				}
+				acc.AddFields("nfs_xprtudp", fields, tags)
+			}
+		}
+	} else if version == "3" || version == "4" {
+		if in(nfs3Fields, first) && len(nline) > 7 {
+			if first == "READ" {
+				fields["read_ops"] = nline[0]
+				fields["read_retrans"] = (nline[1] - nline[0])
+				fields["read_bytes"] = (nline[3] + nline[4])
+				fields["read_rtt"] = nline[6]
+				fields["read_exe"] = nline[7]
+				acc.AddFields("nfsstat_read", fields, tags)
+			} else if first == "WRITE" {
+				fields["write_ops"] = nline[0]
+				fields["write_retrans"] = (nline[1] - nline[0])
+				fields["write_bytes"] = (nline[3] + nline[4])
+				fields["write_rtt"] = nline[6]
+				fields["write_exe"] = nline[7]
+				acc.AddFields("nfsstat_write", fields, tags)
+			}
+		}
+		if fullstat && version == "3" {
+			if in(nfs3Fields, first) && len(nline) <= len(nfsopFields) {
+				for i, t := range nline {
+					item := fmt.Sprintf("%s_%s", first, nfsopFields[i])
```

Comment:
Won't this automatically doubles the cardinality of the measurement?  I have many hosts with hundreds of NFS mounts (both static and autofs), so cardinality issues are a concern.  Many other plugins record read/write metrics as separate metrics, not as tags on a single metric.  For example:
* https://github.com/influxdata/telegraf/blob/master/plugins/inputs/net/net.go#L96
* https://github.com/influxdata/telegraf/blob/master/plugins/inputs/diskio/diskio.go#L141
* https://github.com/influxdata/telegraf/blob/master/plugins/inputs/tomcat/tomcat.go#L172
That said, I'm sure there are counter examples.

Reply:
I don't see how this doubles cardinality. For read/write you get
```
nfs_ops,mountpoint=/mnt/nfs/loki,serverexport=loki:/ READ_bytes_recv=0i,READ_bytes_sent=0i,READ_ops=0i,...,READ_total_time=0i 1550106674000000000
nfs_ops,mountpoint=/mnt/nfs/loki,serverexport=loki:/ WRITE_bytes_recv=0i,WRITE_bytes_sent=0i,WRITE_ops=0i,...,WRITE_total_time=0i 1550106674000000000
```
in your case and 
```
nfs,mountpoint=/mnt/nfs/loki,operation=READ,serverexport=loki:/ bytes_recv=0i,bytes_sent=0i,ops=0i,..,total_time=0i 1550106674000000000
nfs,mountpoint=/mnt/nfs/loki,operation=WRITE,serverexport=loki:/ bytes_recv=0i,bytes_sent=0i,ops=0i,..,total_time=0i 1550106674000000000
```
in danielnelson's example.Both will create two series...

Reply:
> Both will create two series...

I don't think that is correct?  My understanding is that a new series is created from a `measurement`+`tags`.  The two lines of my example have identical tags, and should create 1 series; effectively appending the `READ_*` and `WRITE_*` fields together.

Adding an `operation` for each does create two series, and doubles the cardinality.

```
> insert nfs_ops,mountpoint=/mnt/nfs/loki,serverexport=loki:/ READ_bytes_recv=0i,READ_bytes_sent=0i,READ_ops=0i 1550106674000000000
> insert nfs_ops,mountpoint=/mnt/nfs/loki,serverexport=loki:/ WRITE_bytes_recv=0i,WRITE_bytes_sent=0i,WRITE_ops=0i 1550106674000000000

> show series
key
---
nfs_ops,mountpoint=/mnt/nfs/loki,serverexport=loki:/

> drop measurement nfs_ops

> insert nfs,mountpoint=/mnt/nfs/loki,operation=READ,serverexport=loki:/ bytes_recv=0i,bytes_sent=0i,ops=0i 1550106674000000000
> insert nfs,mountpoint=/mnt/nfs/loki,operation=WRITE,serverexport=loki:/ bytes_recv=0i,bytes_sent=0i,ops=0i 1550106674000000000

> show series
key
---
nfs,mountpoint=/mnt/nfs/loki,operation=READ,serverexport=loki:/
nfs,mountpoint=/mnt/nfs/loki,operation=WRITE,serverexport=loki:/
```

Reply:
I think your understanding is partially correct. In your example you get one series but with the double number of fields. Each line now contains both the `read_*` and `write_*` fields with one of the sets being `null`. That is if you query your series, you will get __all__ line with both fieldsets as you cannot efficiently query by operation. If you, for example, want to query the `read` operation only, you need to check each line if `read_*` is `null` or not. This way visualization or processing becomes either cumbersome or slow.

If on the other hand you stick with multiple measurements as in your later examples the cardinality is exactly the same as if you put the operation as **tag** and write to a **single measurement**.

Citing [the documentation](https://docs.influxdata.com/influxdb/v1.8/concepts/glossary/#series):

> series
> A logical grouping of data defined by shared measurement, tag set, and field key.
>
>series cardinality
>The number of unique database, measurement, tag set, and field key combinations in an InfluxDB instance.

Reply:
This being said, IMO the most discoverable way for a user would be a **single** `nfsstat` measurement with an `operation` tag as you can easily query, process and visualize the operations you are interested in (i.e. you can filter or group-by etc.).

Reply:
Doing this adds a fundamental inconsistency in many of the metrics, which already wasn't great, but we don't need to make it worse.  I've edited this for brevity.  

```
nfsstat,operation=READ ops=700i,retrans=1i,bytes=1407i,rtt=706i,exe=707i 1612648880000000000
nfs_ops, LOOKUP_response_time=306i,LOOKUP_total_time=307i,LOOKUP_ops=300i 1612648880000000000
nfs_bytes normalwritebytes=110857586443i,directreadbytes=783170354688i 1612648880000000000
```
So now there are at least 3 different formats for naming the metrics; it's a mess.

The _wrong_ thing to do here is try to rename any of the fields from what they are canonically called in the source code.  Every time a user space application has tried to "help" do that, it causes confusion and trouble (with `netstat` being the primary offender).

Some of this can be simplified.  I think that `nfs_ops` can be reformatted to use an `operation=` tag.

Reply:
I'm not arguing about the metric name (you could also use `nfs_stat`) but rather that you can make the `operation` a `tag` and have consistent field-names across the operations. Furthermore, I _think_ some queries might be easier (e.g. total operations read and write). if you have the same field-names for read and write.